Sources/Datadog/Core/Utils/ValuePublisher.swift

```diff
@@ -0,0 +1,110 @@
+    /// Type erasure for `ValueObserver` type.
+    private struct AnyObserver<ObservedValue> {
+        let notifyValueChanged: (ObservedValue, ObservedValue) -> Void
+
+        init<Observer: ValueObserver>(wrapped: Observer) where Observer.ObservedValue == ObservedValue {
+            self.notifyValueChanged = wrapped.onValueChanged
+        }
+    }
+
+    /// The queue used to synchronize the access to the `unsafeValue`.
+    /// Concurrent queue is used for performant reads, `.barrier` must be used to make writes exclusive.
+    private let concurrentQueue = DispatchQueue(
+        label: "com.datadoghq.value-publisher-\(type(of: Value.self))",
+        attributes: .concurrent
+    )
+    /// The array of value observers - must be accessed from the `queue`.
+    private var unsafeObservers: [AnyObserver<Value>]
+    /// The managed `Value` - must be accessed from the `queue`.
+    private var unsafeValue: Value {
+        didSet {
+            unsafeObservers.forEach { observer in
+                observer.notifyValueChanged(oldValue, unsafeValue)
+            }
+        }
+    }
+
+    /// The model used for synchronizing `currentValue` updates.
+    enum UpdatesModel {
+        /// The `currentValue` will be updated synchronously, blocking the caller thread.
+        case synchronous
+        /// The `currentValue` will be updated asynchronously, without blocking the caller thread.
+        case asynchronous
+    }
+
+    /// The synchronization model for updating the `unsafeValue`.
+    private let updatesModel: UpdatesModel
+
+    init(initialValue: Value, updatesModel: UpdatesModel) {
+        self.unsafeValue = initialValue
+        self.unsafeObservers = []
+        self.updatesModel = updatesModel
+    }
+
+    /// Registers an observer that will be notified on all value changes.
+    /// All calls to the `observer` will be synchronised using internal concurrent queue.
+    func subscribe<Observer: ValueObserver>(_ observer: Observer) where Observer.ObservedValue == Value {
+        concurrentQueue.async(flags: .barrier) {
+            self.unsafeObservers.append(AnyObserver(wrapped: observer))
+        }
+    }
+
+    /// Registers an observer that will be notified on dictinct value changes.
+    /// All calls to the `observer` will be synchronised using internal concurrent queue.
+    func subscribe<Observer: ValueObserver>(_ observer: Observer) where Observer.ObservedValue == Value, Value: Equatable {
+        concurrentQueue.async(flags: .barrier) {
+            let distinctObserver = DistinctValueObserver(wrapped: observer)
+            self.unsafeObservers.append(AnyObserver(wrapped: distinctObserver))
+        }
+    }
+
+    var currentValue: Value {
+        get {
+            concurrentQueue.sync { unsafeValue }
+        }
+        set {
+            switch updatesModel {
+            case .synchronous:
+                concurrentQueue.sync(flags: .barrier) { unsafeValue = newValue }
```

Comment:
I can't see this problem exactly 💭. The `flags: .barrier` makes the task exclusive - as I understand it can be only executed on the queue if no other task is pending / running. This should guarantee that if a `set` is being called, no `get` will run, no?

Also, this PR has this test:
```swift
    func testValueCanBeWrittenAndReadOnAnyThread() {
         let publisher = ValuePublisher<Int>(initialValue: .mockRandom(), updatesModel: .random())

         // swiftlint:disable opening_brace
         callConcurrently(
             closures: [
                 { publisher.currentValue = .mockRandom() },
                 { _ = publisher.currentValue },
                 { publisher.subscribe(ValueObserverMock<Int>()) }
             ],
             iterations: 50
         )
         // swiftlint:enable opening_brace
     }
```
which should flag any eventual deadlock. Do you see how I can tweak it to reproduce what you describe?

Reply:
It won't let another closure execute but it will make a thread wait for the closure's dequeueing/execution. The setup I have in mind is the following:
```
let queue = DispatchQueue(attributes:.concurrent)

// In some setter:
queue.sync(flags: .barrier) {
    // Do things, call dependencies:
    // ... in some other class that references the previous object (why not):
    // Calling the setter from above will do the following again:
    queue.sync(flags: .barrier) {
         // Will not execute
    }
}
```
This is equivalent to:
```
someSerialQueue.sync {
  // Do things and deadlock
  someSerialQueue.sync {
     // Will not execute
  }
}
```
As well as:
```
let lock = Lock(someNonRecursiveMutex)
// call self recursively...:
let lockAgainAndDeadlock = Lock(someNonRecursiveMutex)
```

This test case setup will not show that kind of situation.
I think you can test this out there by making the `publisher` have a synchronous `updatesModel`, make `ValueObserverMock` keep a reference to the `publisher` and call `publisher.currentValue = .mockRandom()` right in the callback (real world situation will not be so obvious but it can happen).

Reply:
I see 👍. But IMO this doesn't apply here, as the `concurrentQueue` is `private` to `ValuePublisher`. It is impossible to dispatch a task on it from outside. Also, I made sure this runs safely on concurrent threads:
```swift
let publisher = ValuePublisher<Int>(initialValue: .mockRandom())

callConcurrently(
   closures: [
      { publisher.publishSync(publisher.currentValue % 2) },
      { publisher.publishAsync(publisher.currentValue % 2) },
      // ...
   ]
}
```
in https://github.com/DataDog/dd-sdk-ios/pull/422/commits/6d8f233408f855d1cec3a7c0b6ad2dc660788e83.

Of course it's still possible to go into `ValuePublisher's` code and mess it up, but this is mitigated by tests. There's no locking mechanism that is resistant for heedless code changes 🙂.

Reply:
After discussing on Slack, I understood what you mean here 👌. As concluded, I added this note in https://github.com/DataDog/dd-sdk-ios/pull/422/commits/2a20a65294df43e5c155a2970092cad823a369b4:
```swift
/// ...
/// **Note:** as this method is executed on the publisher's queue, call to `publisher.publishSync(_:)`
/// from its implementation will result in deadlock. If the observed value must be updated, use `publishAsync(_:)` instead.
func onValueChanged(oldValue: ObservedValue, newValue: ObservedValue)
```